pkg/process/monitor/process_monitor.go

```diff
@@ -0,0 +1,285 @@
+				return
+			}
+		}
+	}, nil
+}
+
+func (pm *ProcessMonitor) Stop() {
+	close(pm.callbackRunnerDone)
+	close(pm.done)
+	pm.wg.Wait()
+
+	processMonitorLock.Lock()
+	processMonitor = nil
+	processMonitorLock.Unlock()
+}
+
+// GetProcessMonitor() create a monitor (only once) that register to netlink process events.
+//
+// This monitor can monitor.Subscribe(callback, filter) callback on particual event
+// like process EXEC, EXIT. The callback will be called when the filter will match.
+// Filter can be applied on :
+//   process name (NAME)
+//   /proc/pid/smaps file (MAPFILE)
+//   by default ANY is applied
+//
+// Typical initialization:
+// mon := GetProcessMonitor()
+// mon.Subcribe(callback)
+// mon.Initialize()
+//
+// note: o GetProcessMonitor() will always return the same instance
+//         as we can only regiter once with netlink process event
+//       o mon.Subcribe() can be called at anytime, event later
+//
+func GetProcessMonitor() *ProcessMonitor {
+	processMonitorLock.RLock()
+	if processMonitor != nil {
+		defer processMonitorLock.RUnlock()
+		return processMonitor
+	}
+	processMonitorLock.RUnlock()
+
+	processMonitorLock.Lock()
+	defer processMonitorLock.Unlock()
+
+	p := &ProcessMonitor{}
+	p.initializing = true
+	p.events = make(chan netlink.ProcEvent, processMonitorMaxEvents)
+	p.done = make(chan struct{})
+	p.errors = make(chan error)
+	p.procEventCallbacks = make(map[ProcessEventType][]*ProcessCallback)
+
+	if err := netlink.ProcEventMonitor(p.events, p.done, p.errors); err != nil {
+		log.Errorf("could not create process monitor: %s", err)
+		return nil
+	}
+
+	p.callbackRunnerDone = make(chan struct{}, runtime.NumCPU())
+	p.callbackRunner = make(chan func(), runtime.NumCPU())
+	for i := 0; i < runtime.NumCPU(); i++ {
+		p.wg.Add(1)
+		go func() {
+			defer p.wg.Done()
+			for {
+				select {
+				case <-p.callbackRunnerDone:
+					return
+				case call, ok := <-p.callbackRunner:
+					if !ok {
+						continue
+					}
+					call()
+				}
+			}
+		}()
+	}
+
+	// This is the main async loop, where we process processes events from netlink socket
+	// events are dropped until
+	p.wg.Add(1)
+	go func() {
+		defer func() {
+			log.Info("netlink process monitor ended")
+			p.wg.Done()
+		}()
+		for {
+			select {
+			case <-p.done:
+				return
+
+			case event, ok := <-p.events:
+				if !ok {
+					return
+				}
+				p.m.Lock()
+				if p.initializing {
+					p.m.Unlock()
+					continue
+				}
+
+				switch ev := event.Msg.(type) {
+				case *netlink.ExecProcEvent:
+					for _, c := range p.procEventCallbacks[EXEC] {
+						p.evalEXECCallback(c, ev.ProcessPid)
+					}
+				case *netlink.ExitProcEvent:
+					for _, c := range p.procEventCallbacks[EXIT] {
+						p.enqueueCallback(c, ev.ProcessPid)
+					}
+				}
+				p.m.Unlock()
+
+			case err, ok := <-p.errors:
+				if !ok {
+					return
+				}
+				log.Errorf("process montior error: %s", err)
+
+				processMonitorLock.Lock()
+				processMonitor = nil
```

Comment:
why do we want to destroy the singleton?

Reply:
because if got an error from netlink, the socket is closed (should not happen, if we failed here it's mean we are in conflict with something else (outside of datadog)) and recover would mean more complexity as we need to respawn/reconnect/recreate the netlink connection, the underlayer lib is design like that

Reply:
I calling Stop() on netlink lib error